### PR TITLE
Transaction: make signing idiot-proof

### LINKED ISSUE #257: Transaction passes Check() but QbitNinja refuses to broadcast, with error: mandatory-script-verify-flag-failed (Operation not valid with the current stack size) (INVALID)
I'm using the Transaction class, adding some inputs this way:

```
        let transactionDraft = Transaction()
        for inputTx,index,value in inputs do
            transactionDraft.AddInput(inputTx, index) |> ignore
```

And adding some outputs this way:

```
        let dummyAddressForDraftTx = "1KsFhYKLs8qb1GHqrPxHoywNQpet2CtP9t"
        let destAddress = BitcoinAddress.Create(dummyAddressForDraftTx, Network.Main)
        let sumOfInputValues = inputs.Sum(fun (_,_,value) -> value)
        let txOutDraft = TxOut(Money(sumOfInputValues), destAddress)
        transactionDraft.Outputs.Add(txOutDraft)
        Console.WriteLine("num of bytes with draft: " + (transactionDraft.ToBytes().Length.ToString()))
        // then I show fee to user interface, user clicks ACCEPT or CANCEL, then:
        transaction.Outputs.Remove(transaction.Outputs.[0]) |> ignore
        if (transaction.Outputs.Count > 0) then
            failwith "Draft transaction should only contain one output"
        let destAddress = BitcoinAddress.Create(destination, Network.Main)
        let txOutAfterRemovingFee = TxOut(Money(Convert.ToInt64(amount * decimal 100000000) - minerFee.ValueInSatoshis), destAddress)
        transaction.Outputs.Add(txOutAfterRemovingFee)
        // then ask password and recipient to the user
        let encryptedPrivateKey = File.ReadAllText(account.AccountFile.FullName)
        let encryptedSecret = BitcoinEncryptedSecretNoEC(encryptedPrivateKey, Network.Main)
        let privateKey =
            try
                encryptedSecret.GetKey(password)
            with
            | :? SecurityException ->
                raise (InvalidPassword)
        transaction.Sign(privateKey)
```

Then I try to broadcast it this way:

```
        let transactionCheckResult = transaction.Check()
        if (transactionCheckResult <> TransactionCheckResult.Success) then
            failwith (sprintf "Transaction check failed with %s. Please report this bug" (transactionCheckResult.ToString()))
        let client = new QBitNinjaClient(Network.Main)
        let broadcastTask = client.Broadcast(transaction)
        broadcastTask.Wait()
        if not (broadcastTask.Result.Success) then
            failwith (sprintf "Error broadcasting transaction: %s (%s)"
                          broadcastTask.Result.Error.Reason
                          (broadcastTask.Result.Error.ErrorCode.ToString()))
        transaction.GetHash().ToString()
```

And then I get: Error broadcasting transaction: mandatory-script-verify-flag-failed (Operation not valid with the current stack size) (INVALID)

Am I doing something wrong?

### PR DESCRIPTION
**Pull request by knocte**
If transaction.Sign(key) was used, one could have
assumed that he was using an overload in which the
second parameter `bool assumeP2SH` would be optional.
However, there was no overload with a default value
for the second bool parameter; but there is an overload
that takes a `params` keyword for the second parameter,
which means that it can accept an empty array. But being
empty meant that it was not signing at all, even if
the method Sign() was called.

Make it idiot-proof so that if an empty array is passed,
it throws an ArgumentException. On top of that, make
Check() function also check if the transaction is signed.

Fixes #257

### CODE REVIEW

**Comment by NicolasDorier** [comment]
If I merge that then all segwit stuff will break. In segwit, the scriptSig is empty. Even without segwit, it is possible to have an empty scriptSig which is still valid.

Moreover, `Sign` might be done on a transaction which involve multiple parties, one party might sign his part, leaving the rest blank.

If no coin is passed, then no signature should be done, this is how it should be. I have code which will just crash if you check empty array while I was not expecting to crash before.

**Comment by knocte** [comment]
> If I merge that then all segwit stuff will break. In segwit, the scriptSig is empty. Even without segwit, it is possible to have an empty scriptSig which is still valid.

Mmm, ok, so what if we make a TransactionType enum, which is passed to Check() as a parameter? If TransactionType.SegWit is passed, then the Unsigned check doesn't happen.

> If no coin is passed, then no signature should be done, this is how it should be. I have code which will just crash if you check empty array while I was not expecting to crash before.

With `code` you mean external code to NBitcoin or internal? If it's internal, we could make that code use another overload that does the same but doesn't throw on empty array?

You know, I think making the API idiot-proof is important. I even watched your youtube video here last weekend (https://www.youtube.com/watch?v=X4ZwRWIF49w) and I spotted that you even did same initial mistake about using only one argument for Sign() and resulting in an unsigned transaction!

**Comment by NicolasDorier** [comment]
@knocte I mean that if partial signature are expected to work, then I do not expect that if no signature happen there is something thrown. 

>  I spotted that you even did same initial mistake about using only one argument for Sign() and resulting in an unsigned transaction!

I think I should remove the Sign altogether, its behavior (without previous coins passed) is confusing as hell, for me as well. I think I took it from BitcoinJ originally.

`TransactionBuilder.Check()` tells you if the transaction is correctly signed if you pass all the coins.

**Comment by knocte** [comment]
> I think I should remove the Sign altogether, its behavior (without previous coins passed) is confusing as hell, for me as well. I think I took it from BitcoinJ originally.

Sounds good, we mark it as [Obsolete] then? But then we need to supply a message, what to say in it? To use TransactionBuilder?

**Comment by NicolasDorier** [comment]
Yes, or to use `Sign(keys, coins)` instead.

**Comment by knocte** [comment]
> Yes, or to use Sign(keys, coins) instead.

Errrm, what? You just agreed to mark Sign(key) as [Obsolete], but you seem to be forgetting that Sign(key) usage resolves to Sign(key, params coins) where coins is an empty array. You cannot mark something as Obsolete suggesting to use the same thing :)

**Comment by NicolasDorier** [comment]
sorry. What I meant is `Sign(key, bool)` should be obsolete. `Sign(key, params coins)` is how things should be signed.

**Comment by knocte** [comment]
> What I meant is Sign(key, bool) should be obsolete. Sign(key, params coins) is how things should be signed.

That would neither have solved the problem I had the first time I tried to sign a transaction, nor would have solved the confusion you got in your youtube video. Precisely, if both you and me had used `Sign(key,bool)` we would not have had this problem.

**Comment by NicolasDorier** [comment]
`Sign(key, coins)` should not throw. It is designed to try to sign as much as it can given the private keys.

`Sign(key, bool)` assume that you are put in your input's scriptSig the scriptPubKey or redeemScript of the previous output. Such information are actually in the `coins` already. 

By using `Sign(key, coins)` you would support P2SH, P2WSH, P2WPKH, P2PKH, and any combinaison of multi sig.

By using `Sign(key, bool)` you only dumb p2sh and p2pkh, and it ask you to exactract data you already have in your coins, and put the data in the scriptSig before getting called.

In other words, `Sign(key, bool)` requires you more work for supporting less cases.

**Comment by knocte** [comment]
But `Sign(key, coins)` allows beginners shoot themselves in the foot when they don't pass any 2nd argument, thinking they would use default settings to sign.

**Comment by knocte** [comment]
(And not only beginners, as demonstrated by the video ;) )

**Comment by NicolasDorier** [comment]
ha got it. sorry. Maybe I should remove "parms" ?

**Comment by NicolasDorier** [comment]
Yeah indeed you are right, calling `Sign(keys)` make no sense at all. the keyword `params` should be removed.

**Comment by knocte** [comment]
> calling Sign(keys) make no sense at all

Right, that's why I wanted to throw the exception for empty array, as it would throw in this case.

> the keyword params should be removed.

Unfortunately, as far as I understand this would be an API break.

**Comment by NicolasDorier** [comment]
I mean it is possible for one reason or the other that someone pass an empty coins array, because of some logic in their code at runtime, and this is fine, they don't expect it to crash.
However calling `Sign(keys)` is clearly something that never make sense at all. It is a no-op. So it should be prevented at compile time.

**Comment by knocte** [comment]
@NicolasDorier not sure you saw my last message, since you sent one almost exactly at the same time as my last one.

To make sure we're on the same page: removing `params` is desirable, but it would be an API break. Are you willing to do API breaks? I'm guessing you're ok with it so long as you do a major-version change (e.g. from 4.x to 5.x).

As a sidenote, if you're willing to break API, I was thinking this: why not all the way and deprecate non-SegregatedWitness-ways of signing, in this break? This way, we would kill 2 birds with 1 stone.

**Comment by knocte** [comment]
typo: `not all the way` -> `not go all the way`

**Comment by NicolasDorier** [comment]
Non-segwitwitness signature are still widely in use in Bitcoin ecosystem, I guess we are stuck several years before we can deprecate it. For the params change, I am willing to break that. Fixing it is pretty obvious and not much people must depends on it.

**Comment by knocte** [comment]
Closing since this has been address differently in PR#269.

**Comment by knocte** [comment]
typo: address -> addressed

**Comment by NicolasDorier** [comment]
new version pushed on nuget btw

**Comment by knocte** [comment]
Great, we can target many more idiots now :P :)

**Comment by NicolasDorier** [comment]
That's precisely the goal of a well designed library
